File: 삼성기출/[BJ 17142] 연구소3/17142_lab.py
# 7 3
# 2 0 0 0 1 1 0
# 0 0 1 0 1 2 0
# 0 1 1 0 1 0 0
# 0 1 0 0 0 0 0
# 0 0 0 2 0 1 1
# 0 1 0 0 0 0 0
# 2 1 0 0 0 0 2
from itertools import combinations
N, M = map(int, input().split())
arr = []
virus = []
min_time = 9999
dx = [0,0,-1,1] #상하좌우
dy = [-1,1,0,0]

def isIn(nowx, nowy):
    if(0<=nowx<N and 0<=nowy<N and arr[nowy][nowx]!=1):
        return True
    return False
def active_virus(num_list, zero_num):
    global min_time
    visited = [[0]*N for _ in range(N)]
    queue = []
    for i in num_list:
        queue.append([virus[i][0],virus[i][1], 0]) #x, y, time
        visited[virus[i][1]][virus[i][0]] = 0

    max_time = 0
    while(queue):
        vertex = queue.pop(0)
        x = vertex[0]
        y = vertex[1]
        time = vertex[2]
        if([x,y] not in virus):
            max_time = max(max_time, time)
        if (arr[y][x] != 2):
            zero_num -= 1
        if(zero_num == 0):
            break
        for i in range(4):
            nextx = x+dx[i]
            nexty = y+dy[i]
            if(isIn(nextx, nexty) and visited[nexty][nextx] == 0):
                visited[nexty][nextx] = 1
                queue.append([nextx, nexty, time+1])


    #if(check(visited) == True):
    if(zero_num == 0):
        min_time = min(min_time, max_time)

def make_comb(num, prev, num_list):
    if(num==M):
        #바이러스 활성화 함수 호출

        active_virus(num_list)
        return
    for i in range(prev, len(virus)):
        num_list.append(i)
        make_comb(num+1, i+1, num_list)
        num_list.pop(-1)


if __name__ == "__main__":
    zero_num = 0
    num_arr = []
    num = 0
    for i in range(N):
        new_arr = list(map(int, input().split()))
        for j in range(N):
            if(new_arr[j]==2):
                num_arr.append(num)
                num+=1
                virus.append([j, i])# 바이러스 위치 저장
            elif(new_arr[j] == 0):
                zero_num+=1
        arr.append(new_arr)
    #make_comb(0, 0, [])

    comb_list = list(combinations(num_arr, M))
    for i in comb_list:
        active_virus(list(i), zero_num)


    if(min_time == 9999):
        print(-1)
    else:
        print(min_time)

File: 삼성기출/[BJ 17142] 연구소3/17142_lab.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <cstring>
using namespace std;
typedef struct{
    int y, x, t;
}st;
int MAXT = 50*50+1;
int N, M, map[50][50], empty_space;
vector<pair<int, int>> virus;
int dx[4] = {0, 1, 0, -1};
int dy[4] = {-1, 0, 1, 0};
bool isin(int x, int y)
{
    if(0<=x && x<N && 0<=y && y<N) return true;
    return false;
}
int spread_virus(vector<pair<int, int>> &active_virus)
{
    int time = 0, check_empty = empty_space;
    bool visited[50][50];
    bool findflag = false;
    queue<st> q;
    memset(visited, false, sizeof(visited));
    for(auto& v:active_virus){
        q.push({v.first, v.second, 0});
        visited[v.first][v.second] = true;
    }
    
    while(!q.empty())
    {
        st top = q.front();
        q.pop();
        for(int i=0;i<4;i++)
        {
            int nexty = top.y + dy[i];
            int nextx = top.x + dx[i];
            if(isin(nextx, nexty)){
                if(map[nexty][nextx]!=1 && find(active_virus.begin(), active_virus.end(), make_pair(nexty, nextx))==active_virus.end() && !visited[nexty][nextx])
                {
                    time = max(time, top.t+1);
                    q.push({nexty, nextx, top.t+1});
                    visited[nexty][nextx] = true;
                    if(map[nexty][nextx] == 0) check_empty-=1;
                }
                if(check_empty == 0) {
                    findflag = true;
                    break;
                }
            }
        }
        if(findflag) break;
        
    }
    if(findflag) return time;
    else return -1;
}
void choose_virus(vector<pair<int, int>> &active_virus, int index){
    if(active_virus.size() == M){
        int time = spread_virus(active_virus);
        if(time!=-1) MAXT = min(MAXT, time);
        return;
    }
    for(int i=index;i<virus.size();i++){
        active_virus.push_back(virus[i]);
        choose_virus(active_virus, i+1);
        active_virus.pop_back();
    }
    return;
    
}
int main(){
    scanf("%d %d", &N, &M);
    for(int i=0;i<N;i++)
    {
        for(int j=0;j<N;j++)
        {
            cin>>map[i][j];
            if(map[i][j] == 2) virus.push_back(make_pair(i, j));
            else if(map[i][j] == 0) empty_space+=1;
        }
    }
    vector<pair<int, int>> active_virus;
    choose_virus(active_virus, 0);
    if(MAXT == 50*50+1) printf("-1");
    else printf("%d\n", MAXT);
    return 0;
}
